src/Common/FileCache.cpp

```diff
@@ -0,0 +1,623 @@
+{
+    if (fs::exists(cache_base_path))
+        restore();
+    else
+        fs::create_directories(cache_base_path);
+
+    startup_restore_finished = true;
+}
+
+void LRUFileCache::useCell(
+    const FileSegmentCell & cell, FileSegments & result, std::lock_guard<std::mutex> & /* cache_lock */)
+{
+    auto file_segment = cell.file_segment;
+    if (file_segment->download_state == FileSegment::State::DOWNLOADED
+        && fs::file_size(path(file_segment->key(), file_segment->offset())) == 0)
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+                        "Cannot have zero size downloaded file segments. Current file segment: {}",
+                        file_segment->range().toString());
+
+    result.push_back(cell.file_segment);
+
+    /**
+     * A cell receives a queue iterator on first successful space reservation attempt
+     * (space is reserved incrementally on each read buffer nextImpl() call).
+     */
+    if (cell.queue_iterator)
+    {
+        /// Move to the end of the queue. The iterator remains valid.
+        queue.splice(queue.end(), queue, *cell.queue_iterator);
+    }
+}
+
+LRUFileCache::FileSegmentCell * LRUFileCache::getCell(
+    const Key & key, size_t offset, std::lock_guard<std::mutex> & /* cache_lock */)
+{
+    auto it = files.find(key);
+    if (it == files.end())
+        return nullptr;
+
+    auto & offsets = it->second;
+    auto cell_it = offsets.find(offset);
+    if (cell_it == offsets.end())
+        return nullptr;
+
+    return &cell_it->second;
+}
+
+FileSegments LRUFileCache::getImpl(
+    const Key & key, const FileSegment::Range & range, std::lock_guard<std::mutex> & cache_lock)
+{
+    /// Given range = [left, right] and non-overlapping ordered set of file segments,
+    /// find list [segment1, ..., segmentN] of segments which intersect with given range.
+
+    auto it = files.find(key);
+    if (it == files.end())
+        return {};
+
+    const auto & file_segments = it->second;
+    if (file_segments.empty())
+    {
+        removeFileKey(key);
+        return {};
+    }
+
+    FileSegments result;
+    auto segment_it = file_segments.lower_bound(range.left);
+    if (segment_it == file_segments.end())
+    {
+        /// N - last cached segment for given file key, segment{N}.offset < range.left:
+        ///   segment{N}                       segment{N}
+        /// [________                         [_______]
+        ///     [__________]         OR                  [________]
+        ///     ^                                        ^
+        ///     range.left                               range.left
+
+        const auto & cell = (--file_segments.end())->second;
+        if (cell.file_segment->range().right < range.left)
+            return {};
+
+        useCell(cell, result, cache_lock);
+    }
+    else /// segment_it <-- segmment{k}
+    {
+        if (segment_it != file_segments.begin())
+        {
+            const auto & prev_cell = std::prev(segment_it)->second;
```

Comment:
mmm why we need it? `lower_bound` returns iterator to the segment which left boundary is less or equal than required range. Why do we care about previous cell?